src/arithmetic/arithmetic_expression.c

```diff
@@ -578,6 +578,17 @@ bool AR_EXP_IsAttribute(const AR_ExpNode *exp, char **attr) {
 	return true;
 }
 
+bool AR_EXP_ReturnsBoolean(const AR_ExpNode *exp) {
+	ASSERT(exp != NULL && exp->type != AR_EXP_UNKNOWN);
+
+	// If the node does not represent a constant, assume it returns a boolean.
+	// TODO We can add greater introspection in the future if required.
+	if(exp->type == AR_EXP_OP) return true;
```

Comment:
So we assume the expression `1+2` evaluates to boolean?

Reply:
The expression `1+2` would have been reduced to a scalar in arithmetic expression construction, so that would be caught here. If we had an expression like `1+rand()`, that would return true here.

Some invalid queries get caught downstream at https://github.com/RedisGraph/RedisGraph/blob/master/src/execution_plan/optimizations/compact_filters.c#L32, though not the `1+rand()` case, which still does not emit an error but should.

Reply:
This invalid query is now caught in a run-time error.